File: data_structures/tries/trie.py
class Node:
    def __init__(self):
        self.children = {}


class Trie:
    def __init__(self):
        self.root = Node()
    
    # O(K)
    def insert(self, word):
        current_node = self.root
        for char in word:
            if current_node.children.get(char):
                current_node = current_node.children[char]
            else:
                new_node = Node()
                current_node.children[char] = new_node
                current_node = new_node
        current_node.children['*'] = None
        return self

    # O(K)
    def search(self, word):
        current_node = self.root
        for char in word:
            if current_node.children.get(char):
                current_node = current_node.children[char]
            else:
                return None
        return current_node
    
    def _collect_all_words(self, node=None, word='', words=[]):
        current_node = node or self.root
        for char, child_node in current_node.children.items():
            if char == '*':
                words.append(word)
            else:
                self._collect_all_words(child_node, word + char, words)
        return words

    def auto_complete(self, word):
        current_node = self.search(word)
        if not current_node:
            return []
        return self._collect_all_words(current_node)

    def auto_correct(self, word):
        word_found = ''
        current_node = self.root
        for char in word:
            if char in current_node.children:
                word_found += char
                current_node = current_node.children[char]
            else:
                possible_words = self._collect_all_words(current_node)
                if possible_words:
                    return word_found + possible_words[0]
                return possible_words
        return word

File: data_structures/tries/trie.js
class Node {
    constructor() {
        this.children = {};   
    }
}


class Trie {
    constructor() {
        this.root = new Node();
    }

    // O(K)
    insert(word) {
        let currentNode = this.root;
        for (let char of word) {
            if (currentNode.children[char]) {
                currentNode = currentNode.children[char];
            }
            else {
                let newNode = new Node();
                currentNode.children[char] = newNode;
                currentNode = newNode;
            }
        }
        currentNode.children['*'] = null;
        return this;
    }

    
    // O(K)
    search(word) {
        let currentNode = this.root;
        for (let char of word) {
            if (currentNode.children[char]) {
                currentNode = currentNode.children[char];
            }
            else {
                return null;
            }
        }
        return currentNode;
    }
    
    autoComplete(word) {
        let currentNode = this.search(word);
        if (!currentNode) {
            return [];
        }
        return this._collectAllWords(currentNode);
    }


    _collectAllWords(node=null, word='', words=[]) {
        let currentNode = node || this.root;
        for (let key of Object.keys(currentNode.children)) {
            let childNode = currentNode.children[key];
            if (key === '*') {
                words.push(word);
            }
            else {
                this._collectAllWords(childNode, word + key, words);
            }
        }
        return words;
    }

    autoCorrect(word) {
        let wordFound = '';
        let currentNode = this.root;
        for (let char of word) {
            if(currentNode.children[char]) {
                wordFound = wordFound + char;
                currentNode = currentNode.children[char];
            }
            else {
                let possibleWords = this._collectAllWords(currentNode);
                if(possibleWords.length > 0) {
                    return wordFound + possibleWords[0]
                } 
                return possibleWords;
            }
        }
        return word;
    }
}
